File: Day1/knight_problem.py
def is_valid_move(board, row, col, N):
    return row >= 0 and row < N and col >= 0 and col < N and board[row][col] == -1


def knights_tour_util(board, row, col, move_count, N, row_moves, col_moves):
    if move_count == N * N:
        return True

    for i in range(8):
        next_row = row + row_moves[i]
        next_col = col + col_moves[i]
        if is_valid_move(board, next_row, next_col, N):
            board[next_row][next_col] = move_count
            if knights_tour_util(board, next_row, next_col, move_count + 1, N, row_moves, col_moves):
                return True
            board[next_row][next_col] = -1

    return False


def knights_tour(N):
    board = [[-1 for _ in range(N)] for _ in range(N)]
    row_moves = [2, 1, -1, -2, -2, -1, 1, 2]
    col_moves = [1, 2, 2, 1, -1, -2, -2, -1]

    # Start at position (0, 0)
    board[0][0] = 0

    if knights_tour_util(board, 0, 0, 1, N, row_moves, col_moves):
        for row in board:
            print(row)
    else:
        print("No solution exists.")


N = int(input("Enter the value of N, i.e the chess square size: "))
knights_tour(N)

File: Day1/knight_problem.java
import java.util.Arrays;

class KnightTour {
    static final int[] rowMoves = {2, 1, -1, -2, -2, -1, 1, 2};
    static final int[] colMoves = {1, 2, 2, 1, -1, -2, -2, -1};

    static boolean isValidMove(int[][] board, int row, int col, int N) {
        return row >= 0 && row < N && col >= 0 && col < N && board[row][col] == -1;
    }

    static boolean knightsTourUtil(int[][] board, int row, int col, int moveCount, int N) {
        if (moveCount == N * N) {
            return true;
        }

        for (int i = 0; i < 8; i++) {
            int nextRow = row + rowMoves[i];
            int nextCol = col + colMoves[i];
            if (isValidMove(board, nextRow, nextCol, N)) {
                board[nextRow][nextCol] = moveCount;
                if (knightsTourUtil(board, nextRow, nextCol, moveCount + 1, N)) {
                    return true;
                }
                board[nextRow][nextCol] = -1;
            }
        }

        return false;
    }

    static void knightsTour(int N) {
        int[][] board = new int[N][N];
        for (int[] row : board) {
            Arrays.fill(row, -1);
        }

        // Start at position (0, 0)
        board[0][0] = 0;

        if (knightsTourUtil(board, 0, 0, 1, N)) {
            for (int[] row : board) {
                System.out.println(Arrays.toString(row));
            }
        } else {
            System.out.println("No solution exists");
        }
    }

    public static void main(String[] args) {
        int N = 6;
        knightsTour(N);
    }
}

